parcels/basekernel.py

```diff
@@ -0,0 +1,314 @@
+        all_files_array.append(self.log_file)
+        self._cleanup_files = finalize(self, BaseKernel.cleanup_remove_files, self.lib_file, all_files_array, self.delete_cfiles)
+
+    def load_lib(self):
+        self._lib = npct.load_library(self.lib_file, '.')
+        # self._lib = npct.load_library(self.lib_file, get_cache_dir())
+        self._function = self._lib.particle_loop
+        self._cleanup_lib = finalize(self, BaseKernel.cleanup_unload_lib, self._lib)
+
+    def merge(self, kernel, kclass):
+        funcname = self.funcname + kernel.funcname
+        func_ast = None
+        if self.py_ast is not None:
+            func_ast = FunctionDef(name=funcname, args=self.py_ast.args, body=self.py_ast.body + kernel.py_ast.body,
+                                   decorator_list=[], lineno=1, col_offset=0)
+        delete_cfiles = self.delete_cfiles and kernel.delete_cfiles
+        return kclass(self.fieldset, self.ptype, pyfunc=None,
+                      funcname=funcname, funccode=self.funccode + kernel.funccode,
+                      py_ast=func_ast, funcvars=self.funcvars + kernel.funcvars,
+                      c_include=self._c_include + kernel.c_include,
+                      delete_cfiles=delete_cfiles)
+
+    def __add__(self, kernel):
+        if not isinstance(kernel, BaseKernel):
+            kernel = BaseKernel(self.fieldset, self.ptype, pyfunc=kernel)
+        return self.merge(kernel, BaseKernel)
+
+    def __radd__(self, kernel):
+        if not isinstance(kernel, BaseKernel):
+            kernel = BaseKernel(self.fieldset, self.ptype, pyfunc=kernel)
+        return kernel.merge(self, BaseKernel)
+
+    @staticmethod
+    def cleanup_remove_files(lib_file, all_files_array, delete_cfiles):
+        if path.isfile(lib_file):  # and delete_cfiles
+            [remove(s) for s in [lib_file, ] if path.exists(s)]
+        if delete_cfiles and len(all_files_array) > 0:
+            [remove(s) for s in all_files_array if path.exists(s)]
+
+    @staticmethod
+    def cleanup_unload_lib(lib):
+        # Clean-up the in-memory dynamic linked libraries.
+        # This is not really necessary, as these programs are not that large, but with the new random
+        # naming scheme which is required on Windows OS'es to deal with updates to a Parcels' kernel.
+        if lib is not None:
+            try:
+                _ctypes.FreeLibrary(lib._handle) if platform == 'win32' else _ctypes.dlclose(lib._handle)
+            except (OSError, ):
+                logger.warning_once("compiled library already freed.")
+
+    def remove_deleted(self, pset, output_file, endtime):
+        """
+        Utility to remove all particles that signalled deletion.
+
+        This version is generally applicable to all structures and collections
+        """
+        # Indices marked for deletion.
+        indices = [i for i, p in enumerate(pset) if p.state == OperationCode.Delete]
```

Comment:
So this statement could be very slow and inefficient, right? Is the idea not that we force/encourage developers to provide a more efficient way of locating the particles signalled for deletion in the respective datastructure-implementation?

If so, should we just `pass` here, like the other functions below?

Reply:
so, initially, we had 2 incarnations of the `remove_deleted(...)` function: (a) this one, which is a generally-applicable version to remove deleted particles that works for all index-based data structures (e.g. arrays of all sorts), and (b) an SoA-targeted version (called previously `remove_deleted_soa`, see commit history). The SoA-targeted version obviously now remains in `KernelSOA` and _overrides_ this base-class implementation. Essentially: here is a base-implementation that should work for most derivatives - it's not greatly efficient, but it's a working fallback if no more-optimal `remove_deleted(...)` can be developed. Example:

for Array-of-Structure ParticleSets, we do not need to override this function in the derived class, as there is no 'better' option available - slimming down the code and sharing it.

In contrast to the other functions (e.g. `execute`, `exe_python`, `execute_jit`), this function is general enough so it can be shared. Those other functions actually require extensive modifications so a base implementation for them is non-sensical.